feat: adds Content Tagging

pomegranited opened the pull request:
<!--

🌴🌴
🌴🌴🌴🌴         🌴 Note: Palm is in support. Fixes you make on master may still be needed on Palm.
    🌴🌴🌴🌴     If so, make another pull request against the open-release/palm.master branch,
🌴🌴🌴🌴         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌴🌴

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

Adds the models, APIs, and permissions/rules for `openedx.features.content_tagging`, which builds on the `oel_tagging` app changes made by https://github.com/openedx/openedx-learning/pull/62

These changes are the first step towards allow Course Authors to tag content objects (courses, units, libraries) in edx-platform, but none of these changes are user-facing yet.

## Supporting information

Part of https://github.com/openedx/modular-learning/issues/63

## Testing instructions

This change should be 100% covered by unit tests.

But since there's no public-facing APIs yet, if you want to do any further verification, you can test the functionality from the devstack and shell. 

1. Get the `master` [openedx/devstack](https://github.com/openedx/devstack) up and running.
2. Pull this branch into edx-platform
3. Install dependencies and update migrations by running `paver install_prereqs` in the LMS shell (from `devstack` dir, run `make lms-shell`) and CMS shell (`make studio-shell`).
4. Run the migrations added here by running `./manage.py lms migrate` in the LMS shell.
5. Create file `cms/envs/private.py` which contains the following. You may need to restart Studio for these changes to take effect: `make dev.restart-devserver.studio`
   ```
   from .devstack import FEATURES
   # Org-based course authoring restrictions
   FEATURES['ENABLE_CREATOR_GROUP'] = True
   FEATURES['STUDIO_REQUEST_EMAIL'] = 'your@email.com'
   ```
6. Go to the Studio home page and sign up a new user: http://localhost:18010/
7. Expand the "Becoming a Course Creator in Studio" box to request course creation access (this is optional -- an "unrequested" entry is created even if you don't request one, and you can approve that below.)
8. Login to Studio Django Admin in another window as a superuser (http://localhost:18010/admin, `edx@example.com` / `edx`)
9. Create a new Organization, e.g. with Name: OpenCraft, Short Name: OC
10. Navigate to [Course creators|http://localhost:18010/admin/course_creators/coursecreator/] and mark the request `granted`, and select the OpenCraft organization as the only org you can create courses for.

From here, you'll have to play around in the shell, because we only have Django Admin capabilities for ContentTaxonomies right now, and you have to be global staff to access Django Admin, which doesn't let us test out our org restrictions.

For example:

```
# make studio-shell
root@lms:/edx/app/edxapp/edx-platform# ./manage.py cms shell

from django.contrib.auth import get_user_model
# user you created
org_user = get_user_model().objects.get(username='openedx')
# demo users automatically created
learner = get_user_model().objects.get(username='honor')
staff_user = get_user_model().objects.get(username='staff')

from openedx.features.content_tagging import models, rules, api

from organizations.models import Organization
oc = Organization.objects.get(short_name="OC")
org_taxonomy = api.create_taxonomy(name="OpenCraft taxonomy", org_owners=[oc])
org_taxonomy
# <Taxonomy object (2)>
assert org_user.has_perm('oel_tagging.add_taxonomy', org_taxonomy)
assert org_user.has_perm('oel_tagging.change_taxonomy', org_taxonomy)
assert not learner.has_perm('oel_tagging.add_taxonomy')
assert not learner.has_perm('oel_tagging.change_taxonomy', org_taxonomy)
assert not learner.has_perm('oel_tagging.add_taxonomy', org_taxonomy)
```

## Deadline

None

## Other information

Depends on https://github.com/openedx/openedx-learning/pull/62 and https://github.com/openedx/openedx-learning/pull/65

Data migrations added here can be easily rolled back.

Course creator groups added by https://github.com/openedx/edx-platform/pull/26616.

### Author to do before merge:

- [ ] Update dependency: https://github.com/openedx/edx-platform/pull/32661/files#r1253990343

pomegranited committed:
refactor: moves is_content_creator

from cms.djangoapps.contentstore.helpers to common.djangoapps.student.auth

pomegranited committed:
feat: adds content tagging app

Adds models and APIs to support tagging content objects (e.g. XBlocks,
content libraries) by content authors. Content tags can be thought of as
"name:value" fields, though underneath they are a bit more complicated.

* adds dependency on openedx-learning<=0.1.0
* adds tagging app to LMS and CMS
* adds content tagging models, api, rules, admin, and tests.
* content taxonomies and tags can be maintained per organization by
  content creators for that organization.

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited commented on requirements/edx/kernel.in:
```diff
@@ -116,6 +116,9 @@ openedx-django-require
 openedx-filters                     # Open edX Filters from Hooks Extension Framework (OEP-50)
+# FIXME Use pip package once created
+# ref https://github.com/openedx/openedx-learning/pull/62
+git+https://github.com/open-craft/openedx-learning.git@jill/smarter-object-tags-rebase#egg=openedx-learning==0.1
```
To do -- update this dependency when PR merged, tagged, and released on pypi: https://github.com/openedx/openedx-learning/pull/65

ormsbee replied:
Maybe a dummy merge PR openedx-learning to try to force it to push to PyPI?

pomegranited replied:
I worked it out at last.. needed to push a new tag, and that triggered the release :)

https://pypi.org/project/openedx-learning/

ChrisChV commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,163 @@
+        return True
+
+
+class OpenCourseObjectTag(CourseObjectTagMixin, OpenObjectTag):
```
@pomegranited Correct me if I'm wrong, but in this case would only the `valid_for` of `CourseObjectTagMixin` be called, right? The `valid_for` of OpenObjectTag would not be called

pomegranited replied:
Yes.

ChrisChV replied:
In that case, wouldn't it be nice to have both validations called? Since the validations are not being joined, but are being crushed with each other

pomegranited replied:
@ChrisChV You're correct -- it should be calling the subclasses. But python inheritance doesn't work well with class methods, so your comment made me re-examine the `valid_for` and `is_valid` logic.. there was so much duplication there! So I [refactored oel_tagging](https://github.com/open-craft/edx-platform/commit/74a51e203e3a65e191fc91deb5e97616e4d4e7dd) and [the ObjectTag subclasses here](https://github.com/openedx/edx-platform/pull/32661/commits/74a51e203e3a65e191fc91deb5e97616e4d4e7dd) to remove the class method `valid_for`, so we can use normal subclass overrides and super class calls where needed.

Does this work better for you?

ChrisChV replied:
@pomegranited Yes! Looks really good! :+1:

ChrisChV reviewed (approved):
@pomegranited I left a [minor comment](https://github.com/openedx/edx-platform/pull/32661#discussion_r1260189680), but it looks good to me :+1: 

- [X] I tested folowing the testing instructions
- [X] I read through the code and tests with care

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,164 @@
+        if not TaxonomyOrg.is_owner(self.taxonomy, self.org_short_name):
+            return False
+
+        return True
```
Shorthand for this is neater:
```suggestion
        return (
            super()._check_taxonomy() and
            TaxonomyOrg.is_owner(self.taxonomy, self.org_short_name)
        )
```

pomegranited replied:
But I chose to keep it the long way so we can fully ensure test coverage.

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,164 @@
+        if not self.object_key:
+            return False
+
+        return super()._check_object()
```
This is neater:
```suggestion
        return (
            self.object_key and 
            super()._check_object()
        )
```

pomegranited commented on requirements/common_constraints.txt:
```diff
@@ -2,6 +2,22 @@
+# This is a temporary solution to override the real common_constraints.txt
+# In edx-lint, until the pyjwt constraint in edx-lint has been removed.
+# See BOM-2721 for more details.
+# Below is the copied and edited version of common_constraints
```
These changes are an unfortunate artifact of running `make compile_requirements` -- they should be removed.

pomegranited commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,121 @@
+        return org_owners.exists()
+
+
+class ContentTag(ObjectTag):
```
Happy to call this `ContentObjectTag` instead if you think it's clearer?

pomegranited replied:
Done here: https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R69

ChrisChV commented:
@pomegranited :+1: to the new changes

- [X] I tested folowing the testing instructions
- [X] I read through the code and tests with care

bradenmacdonald commented on lms/envs/common.py:
```diff
@@ -3217,6 +3217,10 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
 
+    # Tagging
+    'openedx_tagging.core.tagging.apps.TaggingConfig',
+    'openedx.features.content_tagging',
```
(This can come in a separate PR)

It would be good to add `openedx.features.content_tagging` to [`mypy.ini`](https://github.com/openedx/edx-platform/blob/e7469868207eb3ad614f428a8cbdf7a4b1f48de0/mypy.ini#L9-L14) so you get static type checking, and to [`setup.cfg` `isolated_apps`](https://github.com/openedx/edx-platform/blob/e7469868207eb3ad614f428a8cbdf7a4b1f48de0/setup.cfg#L168-L175) so you can make sure all other parts of the code are only importing from `content_tagging.api`. I think you can add `openedx_tagging.core.tagging` to `isolated_apps` as well to make sure we're only importing from `api`.

pomegranited replied:
Oo that's very cool.. will remember this for a future PR.

pomegranited replied:
cf https://github.com/openedx/modular-learning/issues/83

bradenmacdonald commented on openedx/features/content_tagging/api.py:
```diff
@@ -0,0 +1,135 @@
+        yield ContentTag.cast(object_tag)
+
+
+def tag_object(
```
Maybe the purpose of this wrapper would be more clear if it was named `tag_content_object` ?

pomegranited replied:
Addressed with https://github.com/openedx/edx-platform/pull/32661/commits/ef6d2f9055a8e3bb8a835b6dc0386c41d7420817#diff-0ed9e2727b75df5366745749e3b39718ffbefe4b60289d80feab3b0019feb66bL102-R104.

Also [renamed `get_object_tags` to `get_content_tags`](https://github.com/openedx/edx-platform/pull/32661/commits/ef6d2f9055a8e3bb8a835b6dc0386c41d7420817#diff-0ed9e2727b75df5366745749e3b39718ffbefe4b60289d80feab3b0019feb66bL83-R85).

bradenmacdonald commented on openedx/features/content_tagging/api.py:
```diff
@@ -0,0 +1,135 @@
+def tag_object(
+    taxonomy: Taxonomy,
+    tags: List,
+    object_id: str,
```
I think this would be more consistent with other platform APIs if this took`object_id` as a `UsageKey | CourseKey` (or better, `UsageKey | LearningContextKey`), and then converted it to a string before passing it into `oel_tagging.tag_object`.

pomegranited replied:
Done with https://github.com/openedx/edx-platform/pull/32661/commits/ef6d2f9055a8e3bb8a835b6dc0386c41d7420817#diff-0ed9e2727b75df5366745749e3b39718ffbefe4b60289d80feab3b0019feb66bR107.

bradenmacdonald commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,121 @@
+        """
+        org_owners = cls.get_owners(taxonomy)
+        if not org_owners.exists():
+            return True
```
Is this function `is_enabled` or `is_owner`? Because saying "if nobody is the owner, everyone is the owner" seems weird.

pomegranited replied:
I agree that it's weird.. my only other idea was to add a field to ContentTaxonomy, e.g `available_to_all_orgs`, but I'd have to use a non-proxy model, and those won't cast (because they don't have all the fields).

I also don't want to call this `is_enabled`, because Taxonomy has an `enabled` field..  Do you have any better ideas?

bradenmacdonald replied:
One problem with this current approach is that if an `Organization` is deleted, the `TaxonomyOrg` will be CASCADE deleted too, and then their taxonomy will suddenly go from private to "available for all orgs".

Maybe you can allow `org` to be null, so if there's a `TaxonomyOrg` with `org=None`, that means "make available to all orgs"? Otherwise you'd have to add a `extra_data` JSON field to the base Taxonomy class, which it's better to avoid for now if we can.

pomegranited replied:
Excellent point, thank you! Fixed with [03447f9](https://github.com/openedx/edx-platform/pull/32661/commits/03447f93fb593e1aa2001ffffccb3bf4dc41a215).

bradenmacdonald commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,121 @@
+        Returns the object ID parsed as an Opaque Key, or None if invalid.
+        """
+        if self.object_id:
+            for OpaqueKeyClass in (UsageKey, CourseKey):
```
Shouldn't this use `OPAQUE_KEY_TYPES` which is defined above? Or `(UsageKey, LearningContextKey)` which should cover all the bases.

pomegranited replied:
Oops yep.. missed that in my refactoring..

Fixed with https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R77-R88, though note: instead of `UsageKey`, I need a `BlockUsageLocator` or `LearningContextKey`, because [the TaxonomyOrg check needs `object_key.org`](https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R122).

bradenmacdonald commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,121 @@
+                    return OpaqueKeyClass.from_string(str(self.object_id))
+                except InvalidKeyError:
+                    pass
+        return None
```
Shouldn't this throw an exception rather than silently returning None?

pomegranited replied:
Sure, can do it that way: https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R77-R88

bradenmacdonald commented on openedx/features/content_tagging/models.py:
```diff
@@ -0,0 +1,121 @@
+
+class ContentTaxonomy(Taxonomy):
+    """
+    Taxonomy that accepts ContentTags,
```
_Any_ Taxonomy should "accept" ContentTags since they're a subclass of `ObjecTag`. Might be more clear to say "Taxonomy which can only tag Content objects (e.g. XBlocks or Courses), via ContentTag" (and which ensures a valid TaxonomyOrg owner relationship with the content object.)

pomegranited replied:
Important distinction, thank you.. Updated docstring: https://github.com/openedx/edx-platform/pull/32661/files#diff-8a644422ddd022128d13b6ca7c8b672a33d048d1816a80a80491246bdd76f607R91-R96

bradenmacdonald reviewed (commented):
Looks good, just minor comments. I do hope that we can simplify the casting/validation even more in the future.

bradenmacdonald commented on openedx/features/content_tagging/api.py:
```diff
@@ -0,0 +1,133 @@
+    object_id: Union[BlockUsageLocator, LearningContextKey],
+) -> List[ContentObjectTag]:
+    """
+    Replace the existing ContentObjectTag entries for the given taxonomy + object_id with the given list of tags.
```
I think a slightly higher level of description would be useful in this docstring.
I would suggest including something like:

This is the main API to use when you want to add/update/delete tags from a content object (e.g. an XBlock or course). It works one "Taxonomy" at a time, i.e. one field at a time, so you can set call it with `taxonomy=Keywords, tags=["gravity", "newton"]` to replace any "Keywords" [Taxonomy] tags on the given content object with "gravity" and "newton". Doing so to change the "Keywords" Taxonomy won't affect other Taxonomy's tags (other fields) on the object, such as "Language: [en]" or "Difficulty: [hard]".

... plus the other things you already have in the docstring.

pomegranited replied:
Sure -- I moved away from the "fields" nomenclature in the code because I find it very confusing in context. But I can see why people might want it so I've put your comment in place here: [76d8dcc](https://github.com/openedx/edx-platform/pull/32661/commits/76d8dcca420d45000ea9ae9326c3ab9a6de30f95)

bradenmacdonald replied:
Thanks! I did find this confusing at first, so some wording that like will be helpful I think.

bradenmacdonald commented:
@pomegranited Let me know when the pypi thing is fixed, and I'll merge this for you.

pomegranited commented:
@bradenmacdonald Resolved :) [requirements/edx/kernel.in#L118](https://github.com/openedx/edx-platform/pull/32661/files#diff-1e69f9fed57b4d9a1288ebbd122a504db186ac190d4e1edccd155a024d257029R118)

I also squashed my commits down and rebased on latest `master`. Ready whenever you are, and thank you so much for your guidance here!
